### ISSUE TITLE
ConvTranspose2d made different result on GPU and DLA core in AGX Xavier

### ISSUE THREAD

**Issue by gaizerik** [NVIDIA/TensorRT#1700]
## Environment

**TensorRT Version**: 7.1.3.0 (jetpack4.5.1 default)
**NVIDIA GPU**: AGX Xavier Volta 
**NVIDIA Driver Version**: 
**CUDA Version**: 10.2 (jetpack4.5.1 default)
**CUDNN Version**: 8.0 (jetpack4.5.1 default)
**Operating System**: NVIDIA Jetpack 32.5.1-20210219084708
**Python Version (if applicable)**: 3.6.9
**PyTorch Version (if applicable)**: 1.8.0
** ONNX opset version**: 11

## Description

Hi, I'm trying to infer my model on AGX Xavier's DLA core.

The model works well on the GPU with serialized RT engine.
I've set below option to use DLA core. 
```
            config.set_flag(trt.BuilderFlag.GPU_FALLBACK)
            config.set_flag(trt.BuilderFlag.FP16)
            config.default_device_type = trt.DeviceType.DLA
            config.DLA_core = 0
            config.set_flag(trt.BuilderFlag.STRICT_TYPES)
```
And I got below resulted log(There are no GPU fallbacks).
```
Loading the ONNX file...
Naming the input tensort as "input".
Building the TensorRT engine.  This would take a while...
(Use "--verbose" or "-v" to enable verbose logging.)
Using DLA core 0.
[TensorRT] INFO: 
[TensorRT] INFO: --------------- Layers running on DLA: 
[TensorRT] INFO: {Conv_0,Relu_1,MaxPool_2,Conv_3,Relu_4,Conv_5,Add_6,Relu_7,Conv_8,Relu_9,Conv_10,Add_11,Relu_12,Conv_13,Relu_14,Conv_15,Conv_16,Add_17,Relu_18,Conv_19,Relu_20,Conv_21,Add_22,Relu_23,Conv_24,Relu_25,Conv_26,Conv_27,Add_28,Relu_29,Conv_30,Relu_31,Conv_32,Add_33,Relu_34,Conv_35,Relu_36,Conv_37,Conv_38,Add_39,Relu_40,Conv_41,Relu_42,Conv_43,Add_44,Relu_45,Conv_46,ConvTranspose_47}, 
[TensorRT] INFO: --------------- Layers running on GPU: 
[TensorRT] INFO: 
[TensorRT] INFO: Detected 1 inputs and 2 output network tensors.
Completed creating engine.
```
However, after inference with DLA core, I've got completely different result compare with GPU.
I couldn't find out what was the problem.
So, I subdivided my model and check the output data.

![netron](https://user-images.githubusercontent.com/77386870/147899877-c34d618c-5619-454d-b336-9da21a817ff8.jpeg)

On the above image, all the layers except ConvTranspose are Conv2d, relu and batchnorm only.(It's just a resnet backbone)
I've checked the output0(with ConvTranspose) and output1(just Conv2d).

```
###### With DLA core 0, part of output0 - ConvTranspose result ######
(Pdb) pred0[0,128,0:16, 0:24]
tensor([[0.2228, 0.0115, 0.1920, 0.0115, 0.2325, 0.0115, 0.2017, 0.0115, 0.1979,
         0.0115, 0.1699, 0.0115, 0.1869, 0.0115, 0.1587, 0.0115, 0.1207, 0.0115,
         0.1110, 0.0115, 0.0762, 0.0115, 0.0823, 0.0115],
        [0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115,
         0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115,
         0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115],
        [0.2289, 0.0115, 0.2155, 0.0115, 0.2096, 0.0115, 0.1722, 0.0115, 0.1711,
         0.0115, 0.1783, 0.0115, 0.1990, 0.0115, 0.1849, 0.0115, 0.1674, 0.0115,
         0.1447, 0.0115, 0.1146, 0.0115, 0.0997, 0.0115],
        [0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115,
         0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115,
         0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115],.....

###### With GPU, part of output0 - ConvTranspose result ######
tensor([[ 0.1407,  0.1126,  0.2266,  0.1571,  0.2131,  0.1421,  0.1715,  0.1343,
          0.1823,  0.1969,  0.1368,  0.1735,  0.1454,  0.1672,  0.0878,  0.1201,
          0.1086,  0.1423,  0.1415,  0.1562,  0.0799,  0.1344,  0.0512,  0.1385],
        [ 0.1274,  0.1569,  0.0780,  0.1519,  0.0521,  0.1951,  0.0447,  0.1700,
          0.0862,  0.1626,  0.0478,  0.1371,  0.0614,  0.1610,  0.0519,  0.1368,
          0.0453,  0.1642,  0.0538,  0.1608,  0.0105,  0.1120,  0.0091,  0.1067],
        [ 0.1331,  0.0775,  0.1262,  0.0624,  0.0926,  0.0616,  0.0844,  0.1353,
          0.1547,  0.2013,  0.1028,  0.1643,  0.0777,  0.1606,  0.0496,  0.1168,
          0.0789,  0.1017,  0.0775,  0.1038,  0.0339,  0.1198,  0.0695,  0.1621],
        [ 0.0820,  0.1377,  0.0201,  0.1014,  0.0465,  0.1354,  0.0467,  0.1555,
          0.0921,  0.1420,  0.0510,  0.0923,  0.0045,  0.0800,  0.0635,  0.1190,
          0.0596,  0.1469,  0.0771,  0.0903,  0.0503,  0.0671,  0.0490,  0.0834],......



###### With DLA core 0, part of output1 - Conv2d result ######
tensor([[-0.1051, -0.2180, -0.0660,  0.0415,  0.0630,  0.0908,  0.2247,  0.1699,
          0.1737,  0.1403,  0.0246,  0.0427,  0.0072, -0.0600, -0.0934, -0.0467,
         -0.0028, -0.1235, -0.0560, -0.0687, -0.2073, -0.1119, -0.0832, -0.1705],
        [-0.1259, -0.0532, -0.0922, -0.0822, -0.1218, -0.0112, -0.0615, -0.0740,
         -0.0082,  0.0313, -0.0009,  0.0311,  0.0588,  0.0147, -0.1252, -0.0887,
         -0.1892, -0.1650, -0.0815, -0.1381, -0.1986, -0.1981, -0.1708, -0.2117],
        [-0.2233, -0.1058, -0.1152, -0.0382, -0.0509,  0.0125, -0.0939, -0.2122,
         -0.1044,  0.0588,  0.0212,  0.0320, -0.1053, -0.1714, -0.1703, -0.1512,
         -0.1185, -0.0759, -0.1976, -0.3081, -0.4189, -0.4536, -0.0864, -0.0657],
        [-0.1152, -0.0681, -0.0623,  0.0230,  0.1172,  0.0356, -0.0294, -0.1614,
         -0.1442, -0.0080,  0.0115,  0.1032, -0.0164, -0.0103, -0.1565, -0.2043,
         -0.3037, -0.2023, -0.3315, -0.1445, -0.0603, -0.1133, -0.0987, -0.0996],

###### With GPU, part of output1 - Conv2d result ######
tensor([[-1.0535e-01, -2.1667e-01, -6.5002e-02,  4.5288e-02,  6.3293e-02,
          9.3445e-02,  2.2681e-01,  1.7224e-01,  1.7297e-01,  1.3965e-01,
          2.5360e-02,  4.3335e-02,  6.8359e-03, -6.0364e-02, -9.4666e-02,
         -4.5563e-02,  5.4932e-04, -1.2213e-01, -5.3589e-02, -6.6956e-02,
         -2.0740e-01, -1.1316e-01, -8.2336e-02, -1.7017e-01],
        [-1.2347e-01, -5.0659e-02, -9.3628e-02, -8.5327e-02, -1.2244e-01,
         -1.1108e-02, -6.1218e-02, -7.4707e-02, -7.9956e-03,  3.1891e-02,
         -4.2725e-04,  2.9449e-02,  5.8167e-02,  1.7456e-02, -1.2103e-01,
         -8.6975e-02, -1.8713e-01, -1.6284e-01, -8.2031e-02, -1.3794e-01,
         -1.9885e-01, -2.0020e-01, -1.7285e-01, -2.1204e-01],
        [-2.2241e-01, -1.0583e-01, -1.1530e-01, -3.5126e-02, -4.6875e-02,
          1.1566e-02, -8.9844e-02, -2.1179e-01, -1.0437e-01,  5.8563e-02,
          1.9196e-02,  3.1311e-02, -1.0498e-01, -1.6992e-01, -1.6919e-01,
         -1.5149e-01, -1.1993e-01, -7.4524e-02, -2.0032e-01, -3.1030e-01,
         -4.2090e-01, -4.5361e-01, -8.8501e-02, -6.6467e-02],
        [-1.1432e-01, -6.7810e-02, -6.1096e-02,  2.3605e-02,  1.1676e-01,
          3.4973e-02, -2.6672e-02, -1.5955e-01, -1.4307e-01, -5.2490e-03,
          1.4343e-02,  1.0339e-01, -1.5961e-02, -1.1505e-02, -1.5649e-01,
         -2.0715e-01, -3.0591e-01, -2.0398e-01, -3.2617e-01, -1.4282e-01,
         -5.9906e-02, -1.1304e-01, -9.8022e-02, -1.0046e-01],
```
As you can see, Conv2d result is almost same. The error is less than 0.01.
But for the ConvTranspose2d, seems that it generates sort of garbage data(0.115) for 3 out of 4 pixels in 2x2 kernel.

Is there any know issue like this?

Regards,
Jaden.

**Comment by zerollzeng** [comment]
You can debug accuracy issues with [poligraph](https://github.com/NVIDIA/TensorRT/tree/main/tools/Polygraphy). also see https://github.com/NVIDIA/TensorRT/tree/main/tools/Polygraphy/examples/cli/run/01_comparing_frameworks

I don't recommend marking all layers as outputs since it will break layer fusion in TRT , using the bisecting technique to find the problematic layer is a good practice

**Comment by gaizerik** [comment]
@zerollzeng Thanks for comment!
Even with just one output (convtranspose2d output) I got totally different result between DLA0 and GPU.
The DLA output data itself looks very wired for me.
Looks it has some kind of buffer problem, not accuracy problem because it generates 0.115 periodically as below.
data    0.115
0.115  0.115
### With DLA core 0, part of output0 - ConvTranspose result ###
```
(Pdb) pred0[0,128,0:16, 0:24]
tensor([[0.2228, 0.0115, 0.1920, 0.0115, 0.2325, 0.0115, 0.2017, 0.0115, 0.1979, 0.0115, 0.1699, 0.0115, 0.1869, 0.0115, 0.1587, 0.0115, 0.1207, 0.0115, 0.1110, 0.0115, 0.0762, 0.0115, 0.0823, 0.0115],
        [0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115],
        [0.2289, 0.0115, 0.2155, 0.0115, 0.2096, 0.0115, 0.1722, 0.0115, 0.1711, 0.0115, 0.1783, 0.0115, 0.1990, 0.0115, 0.1849, 0.0115, 0.1674, 0.0115, 0.1447, 0.0115, 0.1146, 0.0115, 0.0997, 0.0115],
        [0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115, 0.0115],.....
```
### With GPU, part of output0 - ConvTranspose result ###
```
tensor([[ 0.1407,  0.1126,  0.2266,  0.1571,  0.2131,  0.1421,  0.1715,  0.1343, 0.1823,  0.1969,  0.1368,  0.1735,  0.1454,  0.1672,  0.0878,  0.1201, 0.1086,  0.1423,  0.1415,  0.1562,  0.0799,  0.1344,  0.0512,  0.1385],
        [ 0.1274,  0.1569,  0.0780,  0.1519,  0.0521,  0.1951,  0.0447,  0.1700, 0.0862,  0.1626,  0.0478,  0.1371,  0.0614,  0.1610,  0.0519,  0.1368, 0.0453,  0.1642,  0.0538,  0.1608,  0.0105,  0.1120,  0.0091,  0.1067],
        [ 0.1331,  0.0775,  0.1262,  0.0624,  0.0926,  0.0616,  0.0844,  0.1353, 0.1547,  0.2013,  0.1028,  0.1643,  0.0777,  0.1606,  0.0496,  0.1168, 0.0789,  0.1017,  0.0775,  0.1038,  0.0339,  0.1198,  0.0695,  0.1621],
        [ 0.0820,  0.1377,  0.0201,  0.1014,  0.0465,  0.1354,  0.0467,  0.1555, 0.0921,  0.1420,  0.0510,  0.0923,  0.0045,  0.0800,  0.0635,  0.1190, 0.0596,  0.1469,  0.0771,  0.0903,  0.0503,  0.0671,  0.0490,  0.0834],......

**Comment by ttyio** [comment]
@gaizerik , have you tried Jetpack 4.6? thanks

**Comment by gaizerik** [comment]
> 

@ttyio , not tried this with JP4.6. I'll try that (It'll take some time).

**Comment by gaizerik** [comment]
> @gaizerik , have you tried Jetpack 4.6? thanks

@ttyio , with Jetpack 4.6,  ConvTranspose of DLA makes nearly same result with that of GPU though I got another issue(GPU fallback related problem. after closing this issue, I'll raise another issue for it).
